### PR TITLE
HDDS-1451 : SCMBlockManager findPipeline and createPipeline are not lock protected.

### PR DESCRIPTION
**Pull request by avijayanhwx**
The getPipelines() and createPipeline() already seem to have a lock in their implementation. However, the problem described here involves a race condition between the call to getPipelines and createPipelines in BlockManagerImpl#allocateBlock. The fix is to add another getPipelines check after a failed createPipeline call to get any newly created pipelines.

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 42 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 415 | trunk passed |
| +1 | compile | 197 | trunk passed |
| +1 | checkstyle | 51 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 883 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 123 | trunk passed |
| 0 | spotbugs | 242 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 425 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 400 | the patch passed |
| +1 | compile | 202 | the patch passed |
| +1 | javac | 202 | the patch passed |
| +1 | checkstyle | 55 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 798 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 135 | the patch passed |
| +1 | findbugs | 472 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 173 | hadoop-hdds in the patch failed. |
| -1 | unit | 1435 | hadoop-ozone in the patch failed. |
| -1 | asflicense | 41 | The patch generated 1 ASF License warnings. |
| | | 5991 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestWatchForCommit |
|   | hadoop.ozone.om.TestOzoneManagerHA |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-799/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/799 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 79fd163f71d1 4.4.0-144-generic #170~14.04.1-Ubuntu SMP Mon Mar 18 15:02:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 7f0e2c6 |
| Default Java | 1.8.0_191 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/1/testReport/ |
| asflicense | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/1/artifact/out/patch-asflicense-problems.txt |
| Max. process+thread count | 5076 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/BlockManagerImpl.java [hunk 1]
```diff
@@ -182,18 +182,28 @@ public AllocatedBlock allocateBlock(final long size, ReplicationType type,
...
         try {
           // TODO: #CLUTIL Remove creation logic when all replication types and
           // factors are handled by pipeline creator
           pipeline = pipelineManager.createPipeline(type, factor);
         } catch (IOException e) {
-          LOG.error("Pipeline creation failed for type:{} factor:{}",
+          LOG.warn("Pipeline creation failed for type:{} factor:{}",
               type, factor, e);
-          break;
+          LOG.info("Checking one more time for suitable pipelines");
```
Comment (mukul1987) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/BlockManagerImpl.java]: Lets add a type and factor here as well.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/BlockManagerImpl.java [hunk 2]
```diff
@@ -182,18 +182,28 @@ public AllocatedBlock allocateBlock(final long size, ReplicationType type,
...
+          availablePipelines = pipelineManager
+              .getPipelines(type, factor, Pipeline.PipelineState.OPEN,
+                  excludeList.getDatanodes(), excludeList.getPipelineIds());
+          if (availablePipelines.size() == 0) {
+            LOG.info("Could not find available pipeline even after trying " +
```
Comment (mukul1987) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/block/BlockManagerImpl.java]: Same as above.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 45 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| -1 | test4tests | 0 | The patch doesn't appear to include any new or modified tests.  Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 559 | trunk passed |
| +1 | compile | 252 | trunk passed |
| +1 | checkstyle | 67 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 1065 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 159 | trunk passed |
| 0 | spotbugs | 294 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 529 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 530 | the patch passed |
| +1 | compile | 253 | the patch passed |
| +1 | javac | 253 | the patch passed |
| +1 | checkstyle | 73 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 794 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 146 | the patch passed |
| +1 | findbugs | 552 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 204 | hadoop-hdds in the patch failed. |
| -1 | unit | 1915 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 48 | The patch does not generate ASF License warnings. |
| | | 7326 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestBlockOutputStream |
|   | hadoop.ozone.TestMiniChaosOzoneCluster |
|   | hadoop.ozone.om.TestOzoneManagerHA |
|   | hadoop.ozone.om.TestOzoneManager |
|   | hadoop.ozone.client.rpc.TestContainerStateMachine |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClientWithRatis |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.client.rpc.TestBCSID |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-799/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/799 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux ecd9e7269767 4.4.0-144-generic #170~14.04.1-Ubuntu SMP Mon Mar 18 15:02:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / c183bd8 |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/2/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/2/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/2/testReport/ |
| Max. process+thread count | 3676 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/server-scm U: hadoop-hdds/server-scm |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-799/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by anuengineer** [comment]
@avijayanhwx  the patch looks good to me. Can you please confirm that the test failures are not due to this patch. Thanks in Advance.

**Comment by avijayanhwx** [comment]
> @avijayanhwx the patch looks good to me. Can you please confirm that the test failures are not due to this patch. Thanks in Advance.

@anuengineer The failures seem unrelated. I have rebased with latest trunk. We can see how this run goes and then commit it.

**Comment by anuengineer** [comment]
+1, LGTM. Thanks.

**Pull request merged**